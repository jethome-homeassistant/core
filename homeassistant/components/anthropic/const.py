"""Constants for the Anthropic integration."""

import logging

DOMAIN = "anthropic"
LOGGER = logging.getLogger(__package__)

DEFAULT_CONVERSATION_NAME = "Claude conversation"

CONF_RECOMMENDED = "recommended"
CONF_PROMPT = "prompt"
CONF_CHAT_MODEL = "chat_model"
RECOMMENDED_CHAT_MODEL = "claude-3-5-haiku-latest"
CONF_MAX_TOKENS = "max_tokens"
RECOMMENDED_MAX_TOKENS = 3000
CONF_TEMPERATURE = "temperature"
RECOMMENDED_TEMPERATURE = 1.0
CONF_THINKING_BUDGET = "thinking_budget"
RECOMMENDED_THINKING_BUDGET = 0
MIN_THINKING_BUDGET = 1024
CONF_WEB_SEARCH = "web_search"
RECOMMENDED_WEB_SEARCH = False
CONF_WEB_SEARCH_USER_LOCATION = "user_location"
RECOMMENDED_WEB_SEARCH_USER_LOCATION = False
CONF_WEB_SEARCH_MAX_USES = "web_search_max_uses"
RECOMMENDED_WEB_SEARCH_MAX_USES = 5
CONF_WEB_SEARCH_CITY = "city"
CONF_WEB_SEARCH_REGION = "region"
CONF_WEB_SEARCH_COUNTRY = "country"
CONF_WEB_SEARCH_TIMEZONE = "timezone"

NON_THINKING_MODELS = [
    "claude-3-5",  # Both sonnet and haiku
    "claude-3-opus",
    "claude-3-haiku",
]

WEB_SEARCH_UNSUPPORTED_MODELS = [
    "claude-3-haiku",
    "claude-3-opus",
    "claude-3-5-sonnet-20240620",
    "claude-3-5-sonnet-20241022",
]
