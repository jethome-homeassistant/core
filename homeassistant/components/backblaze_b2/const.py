"""Constants for the Backblaze B2 integration."""

from collections.abc import Callable
from typing import Final

from homeassistant.util.hass_dict import HassKey

DOMAIN: Final = "backblaze_b2"

CONF_KEY_ID = "key_id"
CONF_APPLICATION_KEY = "application_key"
CONF_BUCKET = "bucket"
CONF_PREFIX = "prefix"

DATA_BACKUP_AGENT_LISTENERS: HassKey[list[Callable[[], None]]] = HassKey(
    f"{DOMAIN}.backup_agent_listeners"
)

METADATA_FILE_SUFFIX = ".metadata.json"
METADATA_VERSION = "1"

BACKBLAZE_REALM = "production"
