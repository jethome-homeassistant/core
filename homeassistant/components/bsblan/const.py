"""Constants for the BSB-Lan integration."""

from __future__ import annotations

from datetime import timedelta
import logging
from typing import Final

# Integration domain
DOMAIN: Final = "bsblan"

LOGGER = logging.getLogger(__package__)
SCAN_INTERVAL = timedelta(seconds=12)  # Legacy interval, kept for compatibility
SCAN_INTERVAL_FAST = timedelta(seconds=12)  # For state/sensor data
SCAN_INTERVAL_SLOW = timedelta(minutes=5)  # For config data

# Services
DATA_BSBLAN_CLIENT: Final = "bsblan_client"

ATTR_TARGET_TEMPERATURE: Final = "target_temperature"
ATTR_INSIDE_TEMPERATURE: Final = "inside_temperature"
ATTR_OUTSIDE_TEMPERATURE: Final = "outside_temperature"

CONF_PASSKEY: Final = "passkey"

DEFAULT_PORT: Final = 80
