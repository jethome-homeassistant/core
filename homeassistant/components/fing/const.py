"""Const for the Fing integration."""

DOMAIN = "fing"
UPNP_AVAILABLE = "upnp_available"
