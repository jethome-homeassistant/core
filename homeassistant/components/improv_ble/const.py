"""Constants for the Improv BLE integration."""

from __future__ import annotations

import asyncio

from homeassistant.util.hass_dict import HassKey

DOMAIN = "improv_ble"

PROVISIONING_FUTURES: HassKey[dict[str, asyncio.Future[str]]] = HassKey(DOMAIN)

# Timeout in seconds to wait for another integration to register a next flow
# after successful provisioning (e.g., ESPHome discovering the device)
PROVISIONING_TIMEOUT = 10.0
