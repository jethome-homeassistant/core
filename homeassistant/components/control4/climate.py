"""Platform for Control4 Climate/Thermostat."""

from __future__ import annotations

from datetime import timedelta
import logging
from typing import Any

from pyControl4.climate import C4Climate
from pyControl4.error_handling import C4Exception

from homeassistant.components.climate import (
    ATTR_TARGET_TEMP_HIGH,
    ATTR_TARGET_TEMP_LOW,
    ClimateEntity,
    ClimateEntityFeature,
    HVACAction,
    HVACMode,
)
from homeassistant.const import ATTR_TEMPERATURE, UnitOfTemperature
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed

from . import Control4ConfigEntry, Control4RuntimeData, get_items_of_category
from .const import CONTROL4_ENTITY_TYPE
from .director_utils import update_variables_for_config_entry
from .entity import Control4Entity

_LOGGER = logging.getLogger(__name__)

CONTROL4_CATEGORY = "comfort"

# Control4 variable names
CONTROL4_HVAC_STATE = "HVAC_STATE"
CONTROL4_HVAC_MODE = "HVAC_MODE"
CONTROL4_CURRENT_TEMPERATURE = "TEMPERATURE_F"
CONTROL4_HUMIDITY = "HUMIDITY"
CONTROL4_COOL_SETPOINT = "COOL_SETPOINT_F"
CONTROL4_HEAT_SETPOINT = "HEAT_SETPOINT_F"

VARIABLES_OF_INTEREST = {
    CONTROL4_HVAC_STATE,
    CONTROL4_HVAC_MODE,
    CONTROL4_CURRENT_TEMPERATURE,
    CONTROL4_HUMIDITY,
    CONTROL4_COOL_SETPOINT,
    CONTROL4_HEAT_SETPOINT,
}

# Map Control4 HVAC modes to Home Assistant
C4_TO_HA_HVAC_MODE = {
    "Off": HVACMode.OFF,
    "Cool": HVACMode.COOL,
    "Heat": HVACMode.HEAT,
    "Auto": HVACMode.HEAT_COOL,
}

HA_TO_C4_HVAC_MODE = {v: k for k, v in C4_TO_HA_HVAC_MODE.items()}

# Map Control4 HVAC state to Home Assistant HVAC action
C4_TO_HA_HVAC_ACTION = {
    "heating": HVACAction.HEATING,
    "cooling": HVACAction.COOLING,
    "idle": HVACAction.IDLE,
    "off": HVACAction.OFF,
}


async def async_setup_entry(
    hass: HomeAssistant,
    entry: Control4ConfigEntry,
    async_add_entities: AddConfigEntryEntitiesCallback,
) -> None:
    """Set up Control4 thermostats from a config entry."""
    runtime_data = entry.runtime_data

    async def async_update_data() -> dict[int, dict[str, Any]]:
        """Fetch data from Control4 director for thermostats."""
        try:
            return await update_variables_for_config_entry(
                hass, entry, VARIABLES_OF_INTEREST
            )
        except C4Exception as err:
            raise UpdateFailed(f"Error communicating with API: {err}") from err

    coordinator = DataUpdateCoordinator[dict[int, dict[str, Any]]](
        hass,
        _LOGGER,
        name="climate",
        update_method=async_update_data,
        update_interval=timedelta(seconds=runtime_data.scan_interval),
        config_entry=entry,
    )

    # Fetch initial data so we have data when entities subscribe
    await coordinator.async_refresh()

    items_of_category = await get_items_of_category(hass, entry, CONTROL4_CATEGORY)
    entity_list = []
    for item in items_of_category:
        try:
            if item["type"] == CONTROL4_ENTITY_TYPE:
                item_name = item["name"]
                item_id = item["id"]
                item_parent_id = item["parentId"]
                item_manufacturer = None
                item_device_name = None
                item_model = None

                for parent_item in items_of_category:
                    if parent_item["id"] == item_parent_id:
                        item_manufacturer = parent_item.get("manufacturer")
                        item_device_name = parent_item.get("roomName")
                        item_model = parent_item.get("model")
            else:
                continue
        except KeyError:
            _LOGGER.exception(
                "Unknown device properties received from Control4: %s",
                item,
            )
            continue

        # Skip if we don't have data for this thermostat
        if item_id not in coordinator.data:
            _LOGGER.warning(
                "Couldn't get climate state data for %s (ID: %s), skipping setup",
                item_name,
                item_id,
            )
            continue

        entity_list.append(
            Control4Climate(
                runtime_data,
                coordinator,
                item_name,
                item_id,
                item_device_name,
                item_manufacturer,
                item_model,
                item_parent_id,
            )
        )

    async_add_entities(entity_list)


class Control4Climate(Control4Entity, ClimateEntity):
    """Control4 climate entity."""

    _attr_has_entity_name = True
    _attr_temperature_unit = UnitOfTemperature.FAHRENHEIT
    _attr_supported_features = (
        ClimateEntityFeature.TARGET_TEMPERATURE
        | ClimateEntityFeature.TARGET_TEMPERATURE_RANGE
        | ClimateEntityFeature.TURN_ON
        | ClimateEntityFeature.TURN_OFF
    )
    _attr_hvac_modes = [HVACMode.OFF, HVACMode.HEAT, HVACMode.COOL, HVACMode.HEAT_COOL]

    def __init__(
        self,
        runtime_data: Control4RuntimeData,
        coordinator: DataUpdateCoordinator[dict[int, dict[str, Any]]],
        name: str,
        idx: int,
        device_name: str | None,
        device_manufacturer: str | None,
        device_model: str | None,
        device_id: int,
    ) -> None:
        """Initialize Control4 climate entity."""
        super().__init__(
            runtime_data,
            coordinator,
            name,
            idx,
            device_name,
            device_manufacturer,
            device_model,
            device_id,
        )

    @property
    def available(self) -> bool:
        """Return if entity is available."""
        return super().available and self._thermostat_data is not None

    def _create_api_object(self) -> C4Climate:
        """Create a pyControl4 device object.

        This exists so the director token used is always the latest one, without needing to re-init the entire entity.
        """
        return C4Climate(self.runtime_data.director, self._idx)

    @property
    def _thermostat_data(self) -> dict[str, Any] | None:
        """Return the thermostat data from the coordinator."""
        return self.coordinator.data.get(self._idx)

    @property
    def current_temperature(self) -> float | None:
        """Return the current temperature."""
        data = self._thermostat_data
        if data is None:
            return None
        return data.get(CONTROL4_CURRENT_TEMPERATURE)

    @property
    def current_humidity(self) -> int | None:
        """Return the current humidity."""
        data = self._thermostat_data
        if data is None:
            return None
        humidity = data.get(CONTROL4_HUMIDITY)
        return int(humidity) if humidity is not None else None

    @property
    def hvac_mode(self) -> HVACMode:
        """Return current HVAC mode."""
        data = self._thermostat_data
        if data is None:
            return HVACMode.OFF
        c4_mode = data.get(CONTROL4_HVAC_MODE) or ""
        return C4_TO_HA_HVAC_MODE.get(c4_mode, HVACMode.OFF)

    @property
    def hvac_action(self) -> HVACAction | None:
        """Return current HVAC action."""
        data = self._thermostat_data
        if data is None:
            return None
        c4_state = data.get(CONTROL4_HVAC_STATE)
        if c4_state is None:
            return None
        # Convert state to lowercase for mapping
        return C4_TO_HA_HVAC_ACTION.get(str(c4_state).lower())

    @property
    def target_temperature(self) -> float | None:
        """Return the target temperature."""
        data = self._thermostat_data
        if data is None:
            return None
        hvac_mode = self.hvac_mode
        if hvac_mode == HVACMode.COOL:
            return data.get(CONTROL4_COOL_SETPOINT)
        if hvac_mode == HVACMode.HEAT:
            return data.get(CONTROL4_HEAT_SETPOINT)
        return None

    @property
    def target_temperature_high(self) -> float | None:
        """Return the high target temperature for auto mode."""
        data = self._thermostat_data
        if data is None:
            return None
        if self.hvac_mode == HVACMode.HEAT_COOL:
            return data.get(CONTROL4_COOL_SETPOINT)
        return None

    @property
    def target_temperature_low(self) -> float | None:
        """Return the low target temperature for auto mode."""
        data = self._thermostat_data
        if data is None:
            return None
        if self.hvac_mode == HVACMode.HEAT_COOL:
            return data.get(CONTROL4_HEAT_SETPOINT)
        return None

    async def async_set_hvac_mode(self, hvac_mode: HVACMode) -> None:
        """Set new target HVAC mode."""
        c4_hvac_mode = HA_TO_C4_HVAC_MODE[hvac_mode]
        c4_climate = self._create_api_object()
        await c4_climate.setHvacMode(c4_hvac_mode)
        await self.coordinator.async_request_refresh()

    async def async_set_temperature(self, **kwargs: Any) -> None:
        """Set new target temperature."""
        c4_climate = self._create_api_object()
        low_temp = kwargs.get(ATTR_TARGET_TEMP_LOW)
        high_temp = kwargs.get(ATTR_TARGET_TEMP_HIGH)
        temp = kwargs.get(ATTR_TEMPERATURE)

        # Handle temperature range for auto mode
        if self.hvac_mode == HVACMode.HEAT_COOL:
            if low_temp is not None:
                await c4_climate.setHeatSetpointF(low_temp)
            if high_temp is not None:
                await c4_climate.setCoolSetpointF(high_temp)
        # Handle single temperature setpoint
        elif temp is not None:
            if self.hvac_mode == HVACMode.COOL:
                await c4_climate.setCoolSetpointF(temp)
            elif self.hvac_mode == HVACMode.HEAT:
                await c4_climate.setHeatSetpointF(temp)

        await self.coordinator.async_request_refresh()
