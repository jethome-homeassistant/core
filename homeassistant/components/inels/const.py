"""Constants for the iNELS integration."""

import logging

from homeassistant.const import Platform

DOMAIN = "inels"
TITLE = "iNELS"

PLATFORMS: list[Platform] = [
    Platform.SWITCH,
]

LOGGER = logging.getLogger(__package__)
