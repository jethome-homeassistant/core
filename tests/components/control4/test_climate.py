"""Test Control4 Climate."""

from unittest.mock import MagicMock

import pytest
from syrupy.assertion import SnapshotAssertion

from homeassistant.components.climate import (
    ATTR_HVAC_MODE,
    ATTR_TARGET_TEMP_HIGH,
    ATTR_TARGET_TEMP_LOW,
    ATTR_TEMPERATURE,
    DOMAIN as CLIMATE_DOMAIN,
    SERVICE_SET_HVAC_MODE,
    SERVICE_SET_TEMPERATURE,
    HVACAction,
    HVACMode,
)
from homeassistant.const import ATTR_ENTITY_ID, Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers import entity_registry as er
from homeassistant.util.unit_system import US_CUSTOMARY_SYSTEM

from . import setup_integration

from tests.common import MockConfigEntry, snapshot_platform

ENTITY_ID = "climate.test_controller_residential_thermostat_v2"


@pytest.fixture
def platforms() -> list[Platform]:
    """Platforms which should be loaded during the test."""
    return [Platform.CLIMATE]


@pytest.fixture
async def init_integration(
    hass: HomeAssistant, mock_config_entry: MockConfigEntry
) -> MockConfigEntry:
    """Set up the Control4 integration for testing."""
    hass.config.units = US_CUSTOMARY_SYSTEM
    await setup_integration(hass, mock_config_entry)
    return mock_config_entry


@pytest.mark.usefixtures(
    "mock_c4_account",
    "mock_c4_director",
    "mock_climate_update_variables",
    "init_integration",
)
async def test_climate_entities(
    hass: HomeAssistant,
    entity_registry: er.EntityRegistry,
    snapshot: SnapshotAssertion,
    mock_config_entry: MockConfigEntry,
) -> None:
    """Test climate entities are set up correctly with proper attributes."""
    await snapshot_platform(hass, entity_registry, snapshot, mock_config_entry.entry_id)


@pytest.mark.parametrize(
    (
        "mock_climate_variables",
        "expected_hvac_mode",
        "expected_hvac_action",
        "expected_temperature",
        "expected_temp_high",
        "expected_temp_low",
    ),
    [
        pytest.param(
            {
                123: {
                    "HVAC_STATE": "off",
                    "HVAC_MODE": "Off",
                    "TEMPERATURE_F": 72.0,
                    "HUMIDITY": 50,
                    "COOL_SETPOINT_F": 75.0,
                    "HEAT_SETPOINT_F": 68.0,
                }
            },
            HVACMode.OFF,
            HVACAction.OFF,
            None,
            None,
            None,
            id="off",
        ),
        pytest.param(
            {
                123: {
                    "HVAC_STATE": "cooling",
                    "HVAC_MODE": "Cool",
                    "TEMPERATURE_F": 74.0,
                    "HUMIDITY": 55,
                    "COOL_SETPOINT_F": 72.0,
                    "HEAT_SETPOINT_F": 68.0,
                }
            },
            HVACMode.COOL,
            HVACAction.COOLING,
            72.0,
            None,
            None,
            id="cool",
        ),
        pytest.param(
            {
                123: {
                    "HVAC_STATE": "heating",
                    "HVAC_MODE": "Auto",
                    "TEMPERATURE_F": 65.0,
                    "HUMIDITY": 40,
                    "COOL_SETPOINT_F": 75.0,
                    "HEAT_SETPOINT_F": 68.0,
                }
            },
            HVACMode.HEAT_COOL,
            HVACAction.HEATING,
            None,
            75.0,
            68.0,
            id="auto",
        ),
    ],
)
@pytest.mark.usefixtures(
    "mock_c4_account",
    "mock_c4_director",
    "mock_climate_update_variables",
    "init_integration",
)
async def test_climate_states(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    expected_hvac_mode: HVACMode,
    expected_hvac_action: HVACAction,
    expected_temperature: float | None,
    expected_temp_high: float | None,
    expected_temp_low: float | None,
) -> None:
    """Test climate entity in different states."""
    state = hass.states.get(ENTITY_ID)
    assert state.state == expected_hvac_mode
    assert state.attributes["hvac_action"] == expected_hvac_action

    assert state.attributes.get("temperature") == expected_temperature

    assert state.attributes.get("target_temp_high") == expected_temp_high
    assert state.attributes.get("target_temp_low") == expected_temp_low


@pytest.mark.parametrize(
    ("hvac_mode", "expected_c4_mode"),
    [
        pytest.param(HVACMode.COOL, "Cool", id="cool"),
        pytest.param(HVACMode.OFF, "Off", id="off"),
    ],
)
@pytest.mark.usefixtures(
    "mock_c4_account",
    "mock_c4_director",
    "mock_climate_update_variables",
    "init_integration",
)
async def test_set_hvac_mode(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    mock_c4_climate: MagicMock,
    hvac_mode: HVACMode,
    expected_c4_mode: str,
) -> None:
    """Test setting HVAC mode."""
    await hass.services.async_call(
        CLIMATE_DOMAIN,
        SERVICE_SET_HVAC_MODE,
        {ATTR_ENTITY_ID: ENTITY_ID, ATTR_HVAC_MODE: hvac_mode},
        blocking=True,
    )
    mock_c4_climate.setHvacMode.assert_called_once_with(expected_c4_mode)


@pytest.mark.parametrize(
    ("mock_climate_variables", "method_name"),
    [
        pytest.param(
            {
                123: {
                    "HVAC_STATE": "idle",
                    "HVAC_MODE": "Heat",
                    "TEMPERATURE_F": 72.5,
                    "HUMIDITY": 45,
                    "COOL_SETPOINT_F": 75.0,
                    "HEAT_SETPOINT_F": 68.0,
                }
            },
            "setHeatSetpointF",
            id="heat",
        ),
        pytest.param(
            {
                123: {
                    "HVAC_STATE": "idle",
                    "HVAC_MODE": "Cool",
                    "TEMPERATURE_F": 74.0,
                    "HUMIDITY": 50,
                    "COOL_SETPOINT_F": 72.0,
                    "HEAT_SETPOINT_F": 68.0,
                }
            },
            "setCoolSetpointF",
            id="cool",
        ),
    ],
)
@pytest.mark.usefixtures(
    "mock_c4_account",
    "mock_c4_director",
    "mock_climate_update_variables",
    "init_integration",
)
async def test_set_temperature(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    mock_c4_climate: MagicMock,
    method_name: str,
) -> None:
    """Test setting temperature in different modes."""
    await hass.services.async_call(
        CLIMATE_DOMAIN,
        SERVICE_SET_TEMPERATURE,
        {ATTR_ENTITY_ID: ENTITY_ID, ATTR_TEMPERATURE: 70.0},
        blocking=True,
    )
    getattr(mock_c4_climate, method_name).assert_called_once_with(70.0)


@pytest.mark.parametrize(
    "mock_climate_variables",
    [
        {
            123: {
                "HVAC_STATE": "idle",
                "HVAC_MODE": "Auto",
                "TEMPERATURE_F": 70.0,
                "HUMIDITY": 50,
                "COOL_SETPOINT_F": 75.0,
                "HEAT_SETPOINT_F": 68.0,
            }
        }
    ],
)
@pytest.mark.usefixtures(
    "mock_c4_account",
    "mock_c4_director",
    "mock_climate_update_variables",
    "init_integration",
)
async def test_set_temperature_range_auto_mode(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    mock_c4_climate: MagicMock,
) -> None:
    """Test setting temperature range in auto mode."""
    await hass.services.async_call(
        CLIMATE_DOMAIN,
        SERVICE_SET_TEMPERATURE,
        {
            ATTR_ENTITY_ID: ENTITY_ID,
            ATTR_TARGET_TEMP_LOW: 65.0,
            ATTR_TARGET_TEMP_HIGH: 78.0,
        },
        blocking=True,
    )
    mock_c4_climate.setHeatSetpointF.assert_called_once_with(65.0)
    mock_c4_climate.setCoolSetpointF.assert_called_once_with(78.0)


@pytest.mark.parametrize("mock_climate_variables", [{}])
@pytest.mark.usefixtures(
    "mock_c4_account",
    "mock_c4_director",
    "mock_climate_update_variables",
    "init_integration",
)
async def test_climate_not_created_when_no_initial_data(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
) -> None:
    """Test climate entity is not created when coordinator has no initial data."""
    # Entity should not be created if there's no data during initial setup
    state = hass.states.get(ENTITY_ID)
    assert state is None


@pytest.mark.parametrize(
    "mock_climate_variables",
    [
        {
            123: {
                "HVAC_STATE": "idle",
                "HVAC_MODE": "Heat",
                # Missing TEMPERATURE_F and HUMIDITY
                "COOL_SETPOINT_F": 75.0,
                "HEAT_SETPOINT_F": 68.0,
            }
        }
    ],
)
@pytest.mark.usefixtures(
    "mock_c4_account",
    "mock_c4_director",
    "mock_climate_update_variables",
    "init_integration",
)
async def test_climate_missing_variables(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
) -> None:
    """Test climate entity handles missing variables gracefully."""
    state = hass.states.get(ENTITY_ID)
    assert state is not None
    assert state.state == HVACMode.HEAT
    assert state.attributes.get("current_temperature") is None
    assert state.attributes.get("current_humidity") is None
    assert state.attributes["temperature"] == 68.0


@pytest.mark.parametrize(
    "mock_climate_variables",
    [
        {
            123: {
                "HVAC_STATE": "idle",
                "HVAC_MODE": "UnknownMode",
                "TEMPERATURE_F": 72.0,
                "HUMIDITY": 50,
                "COOL_SETPOINT_F": 75.0,
                "HEAT_SETPOINT_F": 68.0,
            }
        }
    ],
)
@pytest.mark.usefixtures(
    "mock_c4_account",
    "mock_c4_director",
    "mock_climate_update_variables",
    "init_integration",
)
async def test_climate_unknown_hvac_mode(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
) -> None:
    """Test climate entity handles unknown HVAC mode."""
    state = hass.states.get(ENTITY_ID)
    assert state is not None
    assert state.state == HVACMode.OFF  # Defaults to OFF for unknown modes


@pytest.mark.parametrize(
    "mock_climate_variables",
    [
        {
            123: {
                "HVAC_STATE": "unknown_state",
                "HVAC_MODE": "Heat",
                "TEMPERATURE_F": 72.0,
                "HUMIDITY": 50,
                "COOL_SETPOINT_F": 75.0,
                "HEAT_SETPOINT_F": 68.0,
            }
        }
    ],
)
@pytest.mark.usefixtures(
    "mock_c4_account",
    "mock_c4_director",
    "mock_climate_update_variables",
    "init_integration",
)
async def test_climate_unknown_hvac_state(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
) -> None:
    """Test climate entity handles unknown HVAC state."""
    state = hass.states.get(ENTITY_ID)
    assert state is not None
    assert state.attributes.get("hvac_action") is None
