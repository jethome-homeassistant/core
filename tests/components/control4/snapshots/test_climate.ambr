# serializer version: 1
# name: test_climate_entities[climate.test_controller_residential_thermostat_v2-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'hvac_modes': list([
        <HVACMode.OFF: 'off'>,
        <HVACMode.HEAT: 'heat'>,
        <HVACMode.COOL: 'cool'>,
        <HVACMode.HEAT_COOL: 'heat_cool'>,
      ]),
      'max_temp': 95,
      'min_temp': 45,
    }),
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'climate',
    'entity_category': None,
    'entity_id': 'climate.test_controller_residential_thermostat_v2',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Residential Thermostat V2',
    'platform': 'control4',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': <ClimateEntityFeature: 387>,
    'translation_key': None,
    'unique_id': '123',
    'unit_of_measurement': None,
  })
# ---
# name: test_climate_entities[climate.test_controller_residential_thermostat_v2-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'current_humidity': 45,
      'current_temperature': 72,
      'friendly_name': 'Test Controller Residential Thermostat V2',
      'hvac_action': <HVACAction.IDLE: 'idle'>,
      'hvac_modes': list([
        <HVACMode.OFF: 'off'>,
        <HVACMode.HEAT: 'heat'>,
        <HVACMode.COOL: 'cool'>,
        <HVACMode.HEAT_COOL: 'heat_cool'>,
      ]),
      'max_temp': 95,
      'min_temp': 45,
      'supported_features': <ClimateEntityFeature: 387>,
      'target_temp_high': None,
      'target_temp_low': None,
      'temperature': 68,
    }),
    'context': <ANY>,
    'entity_id': 'climate.test_controller_residential_thermostat_v2',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'heat',
  })
# ---
